src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
     errors.into_result(())
 }
 
+// Go through all functions and check if they are castable
+// based only on their args
+pub fn initialize_castable(functions: &mut FunctionMap, types: &TypeMap) {
+    for func in functions.values_mut() {
+        for arg in &func.args {
+            if arg.param_type.is_associated_resource(types) {
+                func.is_castable = false;
+
+                // If we have found one associated resource
+                // we can continue
+                continue;
+            }
+        }
+    }
+}
+
+// Go through all of the functions and check if they are castable
+// base on functions they call.
+pub fn determine_castable(functions: &mut FunctionMap, types: &TypeMap) -> u64 {
+    let mut num_changed: u64 = 0;
+    // We need tmp_functions to avoid a immutable borrow after a mutable one.
+    let tmp_functions = functions.clone();
+    'outer: for func in functions.values_mut() {
+        // If we are already false there is no reason to check our called functions.
+        if !func.is_castable {
+            continue;
+        }
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                if let Some(inner_func) = tmp_functions.get(&call.get_cil_name()) {
+                    if !inner_func.is_castable {
+                        num_changed += 1;
+                        func.is_castable = false;
+                        continue 'outer;
+                    }
+                // The call may be a built-in so check the args
+                } else {
+                    for arg in &call.args {
+                        if let Argument::Var(arg) = &arg.0 {
+                            // Need to special case this.*
+                            if arg.to_string().contains("this.") {
+                                num_changed += 1;
+                                func.is_castable = false;
+                                continue 'outer;
+                            }
+                            if let Some(ti) = types.get(arg.as_ref()) {
+                                if ti.is_associated_resource(types) {
+                                    num_changed += 1;
+                                    func.is_castable = false;
+                                    continue 'outer;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    num_changed
+}
+
+pub fn initialize_terminated<'a>(
+    functions: &'a FunctionMap<'a>,
+) -> (Vec<String>, Vec<FunctionInfo<'a>>) {
+    let mut term_ret_vec: Vec<String> = Vec::new();
+    let mut nonterm_ret_vec: Vec<FunctionInfo> = Vec::new();
+
+    for func in functions.values() {
+        let mut is_term = true;
+
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
```

Comment:
What about calls in other kinds of statements (optional blocks, if blocks etc)?

Something like this is non-terminating:

```
fn foo(...) {
    if (true) {
        something.some_call(...);
    }
}
```

Reply:
If I am iterating through the original_body like I am will I see the something.some_call or do I have to recursively go through the IfBlock?

Reply:
You have to recursively go into the blocks.  The Statement in the original_body will just be IfBlock or whatever, and then that block will contain other statements (IIRC, IfBlock doesn't yet, but will eventually.  Optional should already I believe)

Reply:
We should probably add some sort of helper to handle the block traversal logic.  A function that takes a Vec<Statement> and returns an iterator through all the statements descending into the blocks, defined in ast.rs would be really helpful.

Reply:
I already started doing this a slightly different way before I saw this comment.  I will upload shortly but can change the approach if you think necessary.